pkg/aws/awscontrolplane/validate_awscontrolplane.go

```diff
@@ -43,18 +43,98 @@ func NewValidator(config config.Config) (*Validator, error) {
 
 func (v *Validator) Validate(request *v1beta1.AdmissionRequest) (bool, error) {
 	var awsControlPlane infrastructurev1alpha2.AWSControlPlane
-	var allowed bool
 
 	if _, _, err := validator.Deserializer.Decode(request.Object.Raw, nil, &awsControlPlane); err != nil {
 		return false, microerror.Maskf(aws.ParsingFailedError, "unable to parse awscontrol plane: %v", err)
 	}
-	allowed, err := v.ControlPlaneLabelMatch(awsControlPlane)
+	controlPlaneLabelMatches, err := v.ControlPlaneLabelMatch(awsControlPlane)
 	if err != nil {
 		return false, microerror.Mask(err)
 
 	}
+	azCountAllowed, err := v.AZCount(awsControlPlane)
+	if err != nil {
+		return false, microerror.Mask(err)
+
+	}
+	azReplicaMatches, err := v.AZReplicaMatch(awsControlPlane)
+	if err != nil {
+		return false, microerror.Mask(err)
+
+	}
+
+	return controlPlaneLabelMatches && azCountAllowed && azReplicaMatches, nil
+}
+
+func (v *Validator) AZReplicaMatch(awsControlPlane infrastructurev1alpha2.AWSControlPlane) (bool, error) {
+	var g8sControlPlane infrastructurev1alpha2.G8sControlPlane
+	var err error
+	var fetch func() error
+
+	// Fetch the G8sControlPlane.
+	{
+		v.Log("level", "debug", "message", fmt.Sprintf("Fetching G8sControlPlane %s", awsControlPlane.Name))
+		fetch = func() error {
+			ctx := context.Background()
+
+			err = v.k8sClient.CtrlClient().Get(
+				ctx,
+				types.NamespacedName{Name: awsControlPlane.GetName(), Namespace: awsControlPlane.GetNamespace()},
+				&g8sControlPlane,
+			)
+			if err != nil {
+				return microerror.Maskf(aws.NotFoundError, "failed to fetch g8sControlplane: %v", err)
```

Comment:
FMI how does the error handling work here? Why is there some public error in some other package?

Reply:
We define a few errors and function that all handlers share. I think the package is called aws because we used to have both azure and aws.

Reply:
Yes, that explains what we do. I am just wondering why we do it that way. Based on my experience it is best practice to keep errors to the actual package scope and thus keep them private. We cover quite some information in https://github.com/giantswarm/fmt/blob/master/go/errors.md.

Reply:
Yeah that makes sense to me, too. I changed the error handling for the `awscontrolplane` and `g8scontrolplane` packages.